ddl/index.go

```diff
@@ -1046,21 +1116,20 @@ func (w *addIndexWorker) batchCheckUniqueKey(txn kv.Transaction, idxRecords []*i
 	for i, key := range w.batchCheckKeys {
 		if val, found := batchVals[string(key)]; found {
 			if w.distinctCheckFlags[i] {
-				handle, err1 := tablecodec.DecodeHandleInUniqueIndexValue(val, w.table.Meta().IsCommonHandle)
-				if err1 != nil {
-					return errors.Trace(err1)
-				}
-
-				if handle != idxRecords[i].handle {
```

Comment:
This line is buggy because it is trying to use reference equality to check if the index key is duplicated. The `handle` returned by `DecodeHandleInUniqueIndexValue` would never equal to `idxRecords[i].handle`, which means the index will always be treated as a duplicate.

I fixed this potential critical bug as part of the rewrite of the decoding of the index key-value.

Reply:
The good news is we haven't done the same thing in another place. I have checked the codebase at `master` using this naive analyzer.

```go
func run(pass *analysis.Pass) (interface{}, error) {
	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
	nodeFilter := []ast.Node{
		(*ast.BinaryExpr)(nil),
	}

	const typeName = "github.com/pingcap/tidb/kv.Handle"

	inspect.Preorder(nodeFilter, func(node ast.Node) {
		expr := node.(*ast.BinaryExpr)

		if expr.Op != token.EQL && expr.Op != token.NEQ {
			return
		}

		if pass.TypesInfo.TypeOf(expr.X).String() != typeName || pass.TypesInfo.TypeOf(expr.Y).String() != typeName {
			return
		}

		pass.Reportf(expr.Pos(), "use %s on two handle type", expr.Op.String())
	})

	return nil, nil
}
```

The output contains only one line and it says
```
tidb/ddl/index.go:1054:8: use != on two handle type
```

Reply:
Should we add this static check to `make check`?